### DISCUSSION TITLE
Trigger an extension from terminal command.

### DISCUSSION THREAD

**Discussion by krishnagogada** [eclipse-theia/theia#9084] [Q&A]
Is it possible to trigger a extension when we enter the specific command?

I wrote an extension which trigger when command alt+h.

Similarly in terminal when I enter `start my-project` then particular extension should be trigger.

Thanks in advance for your help.

**Comment by vince-fugnitto** [comment]
@krishnagogada thank you for the discussion, do you mind providing additional details so the community can help you.

- What do you mean when you say "trigger an extension"?
- Are you creating a custom `@theia` extension or vscode plugin (if a vscode plugin you will need to [activationEvents](https://code.visualstudio.com/api/references/activation-events)).

**Reply by krishnagogada** [reply]
@vince-fugnitto thank for replying.

1. Trigger an extension means execute an extension.
2. I am created a custom `@theia` extension.

What my use case is when i entered a command in `@theia/terminal` (Example :- `start my-project`) then execute an extension(which is my custom `@theia` extension) .

Is it possible, to execute an custom extension in this way?

**Reply by paul-marechal** [reply]
The terminal relays what you type to a real shell. Shells usually spawn actual processes. From there I see two ways:

1. Make a script/program named `start` that would somehow communicate with the Theia instance process and trigger your extension via this IPC.
2. Somehow find how to intercept what is being sent over to the shell to recognize your custom command, but this sounds a bit complex to properly do.

**Reply by krishnagogada** [reply]
@marechal-p i am really new to this theia. I got some questions.
1. Make a script/program named `start` means we create a script with named `start` in **package.json** and communicate with the extension.  Then in which **package.json** we have to keep, in extension or in browser.
2. I didn't get what IPC means.
3. I too have the same idea that to recognise my custom command and trigger the extension. But how to recognise the command, means is there any methods to get the current entered command.

**Reply by paul-marechal** [reply]
1. I meant something that can be spawned as a process, an executable file, a script to run with node, a script from `package.json`... When you type `start a b c<enter>` in Theia's terminal, those characters are directly forwarded to a shell, so the shell will do shell things and look for a `start` command to run. So make a `start` command available in any way you prefer. Doesn't matter which. This is not Theia specific, just regular OS things.
2. Interprocess communication. Since the shell will start a new process, you need to setup a way to communicate back with Theia's process to do things with your extension. Again, you'll have to figure out the details on your own.
3. If you want to intercept characters you should look into the `TerminalProcess` and/or `ShellProcess` classes and see how to rebind those or the mechanism around it to do what you want. I don't know precisely what to change, but look for the place where the typed characters are flowing through.

**Reply by krishnagogada** [reply]
Similar doubt is it possible to open a folder using command in theia terminal for example `open my-project`?

**Reply by paul-marechal** [reply]
Nothing like such is currently implemented in the framework, but some products (e.g. Gitpod) support it.
